### DISCUSSION TITLE
PID on weighted average of BT and ET?

### DISCUSSION THREAD

**Discussion by russ999** [artisan-roaster-scope/artisan#998] [Q&A]
I see PID can be based on ET or BT. How about a weighted average of both?  I have a popper style air roaster. My ET probe is near the heater below my beans. BT probe is in the air above the beans. So I assume an accurate temp is somewhere in between (though not perfectly in the middle). The only time it's an issue is if I roast different quantities, and the spread between ET and BT varies a lot. Large quantity of beans, large spread.  Small quantity, small spread.  I assume profile roasting would be more consistent if I base my PID on a blend of the two temperatures. I'm guessing 70% BT / 30% ET. How can I set this up on artisan?  Is there a code snippet I can stick in somewhere? Maybe on the arduino as it reports temperatures?  

I run artisan on a MacBook, with arduino and a TC4 running aArtisanQ_PID.

**Comment by bigl00z3** [comment]
Not sure how it will work on the PID regulation. 

But to answer your question on how to use the control based on both values in Artisan, you could:

-  Have the native ET and BT set in `Config/Device.../Extra Devices` here an example using a Phidget input. Just select your device from the drop down menu
![image](https://user-images.githubusercontent.com/8886922/200032836-66f054cc-defa-4c76-bbe1-e8217129cd90.png)
If you still want to see the original values, just tick the LCD and curve boxes

- Then in `Config/Device.../Symb ET/BT` you make a calculation in between your 2 Extra Devices (as it is on the first line they are Y1 and Y2 symbolic variable). In the below calculation example, I just average in between the 2 values, but you decide what calculation you prefer:
![image](https://user-images.githubusercontent.com/8886922/200034368-0c9eac3c-6d45-4b6e-8ca1-94abdc43966a.png)

- In the PID Control window, you use the new calculated BT as the input for the PID:
![image](https://user-images.githubusercontent.com/8886922/200033927-dc81ba66-914f-4644-a887-3a0a9d1c1448.png)

Nico

**Reply by MAKOMO** [reply]
As you are running the (to Artisan external) TC4 PID, this averaging has to be done in the TC4 firmware. If you instead use the Artisan internal software PID (which is not advisable as it runs at a slower speed then the TC4 one) you can use symbolic assignments to modify the BT or ET signal as Nico proposed to achieve this effect or wait for the next Artisan release which will allow you also to choose one of the extra device curves as PID source (you still need to use symbolic assignments to fill this extra device curve with the average you are after).
